# Faiss vector storage

In [3]:
import pandas as pd
pd.set_option('display.max_colwidth', 100)

df = pd.read_csv("sample_text.csv")
df

text  \
0                                        Meditation and yoga can improve mental health   
1                     Fruits, whole grains and vegetables helps control blood pressure   
2                                    These are the latest fashion trends for this week   
3                                    Vibrant color jeans for male are becoming a trend   
4                                                   The concert starts at 7 PM tonight   
5                      Navaratri dandiya program at Expo center in Mumbai this october   
6                                    Exciting vacation destinations for your next trip   
7  Maldives and Srilanka are gaining popularity in terms of low budget vacation places   

  category  
0   Health  
1   Health  
2  Fashion  
3  Fashion  
4    Event  
5    Event  
6   Travel  
7   Travel

In [4]:
from sentence_transformers import SentenceTransformer
encoder = SentenceTransformer("all-mpnet-base-v2") # model
vectors = encoder.encode(df.text)
print(vectors.shape)

c:\Users\sebat\Proyects\equity_research_tool\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\sebat\Proyects\equity_research_tool\venv\Lib\site-packages\huggingface_hub\file_download.py:140: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\sebat\.cache\huggingface\hub\models--sentence-transformers--all-mpnet-base-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python 

(8, 768)


Saving the first vector into 'dim'

In [5]:
dim = vectors.shape[1]
dim

768

Faiss creates an index that will improve and make much faster index searches later on 

In [6]:
import faiss
index = faiss.IndexFlatL2(dim)
index

<faiss.swigfaiss.IndexFlatL2; proxy of <Swig Object of type 'faiss::IndexFlatL2 *' at 0x0000028EF7ED2760> >

In [7]:
index.add(vectors)

In [ ]:
search_query = "I want to buy a Polo T-Shirt"
vec = encoder.encode(search_query) # encode the query
vec.shape

(768,)

In [13]:
# Since daiss needs a 2 dim array, we will use numpy 
import numpy as np
vec_np = np.array(vec).reshape(1, -1)
vec_np.shape

(1, 768)

In [15]:
index.search(vec_np, k=2) # k parameters is how many similar vectors do you want, for eg. 2 similar vectors

(array([[1.3844838, 1.4039092]], dtype=float32), array([[3, 2]], dtype=int64))

the first output is the distances : array([[1.3844838, 1.4039092]]

the second one is the index in the original df: array([[3, 2]]

In [16]:
distances, indexes = index.search(vec_np, k=2)

In [20]:
indexes[0]

array([3, 2], dtype=int64)

In [21]:
df.loc[indexes[0]]

text category
3  Vibrant color jeans for male are becoming a trend  Fashion
2  These are the latest fashion trends for this week  Fashion